### PR TITLE
Better consumer offset handling

### PR DESCRIPTION
**Pull request by wvanbergen**
- Add `Client.GetOffsetRange` to get the available range of offsets for a partition in one request. I had to make some changes to `OffsetRequest` and `OffsetResponse` to make it possible to ask for more than one offset per partition, but no changes to the public API were necessary.

- When calling ConsumePartition, always check whether the offset is within the offset range. This means we now have to do an OffsetRequest for every `ConsumePartition` call, even if the offset is provided. The good news is that the method will immediately return an error and never start a goroutine, instead of starting the goroutine and returning an error in the `Errors()` channel which you can easily ignore.

@Shopify/kafka

### CODE REVIEW

**Comment by eapache** [comment]
> make it possible to ask for more than one offset per partition

Is that even possible? It's never explicitly stated, but I'd always assumed based on the layout of the protocol that it wasn't really supported to include multiple blocks for a single partition.

**Comment by wvanbergen** [comment]
Apparently it works in Kafka 0.8.1 and 0.8.2. The OffsetRequest documentation is especially unclear, but this led me to believe that is was supported:

> The response contains the starting offset of each segment for the requested partition as well as the "log end offset" i.e. the offset of the next message that would be appended to the given partition.

You can even do more than two, as long as you request the timestamps in descending order.

File: client.go [hunk 1]
```diff
@@ -292,6 +297,42 @@ func (client *client) RefreshMetadata(topics ...string) error {
 	return client.tryRefreshMetadata(topics, client.conf.Metadata.Retry.Max)
 }
 
+func (client *client) GetOffsetRange(topic string, partitionID int32) (int64, int64, error) {
```
Comment (eapache) [inline client.go]: Why do all this work when you could have just made two `GetOffset` calls?

Comment (wvanbergen) [inline client.go]: 1 request vs. 2 requests.

File: offset_request.go [hunk 2]
```diff
@@ -12,7 +12,7 @@ func (r *offsetRequestBlock) encode(pe packetEncoder) error {
 }
 
 type OffsetRequest struct {
-	blocks map[string]map[int32]*offsetRequestBlock
+	blocks map[string]map[int32][]*offsetRequestBlock
```
Comment (eapache) [inline offset_request.go]: since there's no need for uniqueness per partition anymore, this can just become `map[string][]*offsetRequestBlock` and you can put a field `partition int32` into `offsetRequestBlock`.

Comment (eapache) [inline offset_request.go]: Since this is apparently legit (who knew) we should probably update all the other request types as well - perhaps that deserves a separate PR before this one.

File: client.go (hunk 1 above)
Comment (eapache) [inline client.go]: it only happens once on startup, doesn't seem like a big deal (premature optimization and all that)

Comment (wvanbergen) [inline client.go]: In general I find this something that is generally useful. E.g., for monitoring purposes, you'd also want to know the available range. Of course, we could do two requests for this purpose as well.

File: offset_request.go [hunk 3]
```diff
@@ -51,16 +53,20 @@ func (r *OffsetRequest) version() int16 {
...
+		r.blocks = make(map[string]map[int32][]*offsetRequestBlock)
 	}
 
 	if r.blocks[topic] == nil {
-		r.blocks[topic] = make(map[int32]*offsetRequestBlock)
+		r.blocks[topic] = make(map[int32][]*offsetRequestBlock)
+	}
+
+	if r.blocks[topic][partitionID] == nil {
+		r.blocks[topic][partitionID] = make([]*offsetRequestBlock, 0, 1)
```
Comment (eapache) [inline offset_request.go]: this is unnecessary, `nil` is a valid thing to pass to `append`

Comment (wvanbergen) [inline offset_request.go]: TIL

**Comment by wvanbergen** [comment]
WTF :/ Github marked this as merged after I rebased.

**Comment by wvanbergen** [comment]
Closing in favor of #418
